src/features/Folding.ts

```diff
@@ -0,0 +1,601 @@
+     * @param logger The logging object to send messages to
+     * @returns      A grammar parser for the PowerShell language is succesful or undefined if an error occured
+     */
+    public grammar(logger: Logger): IGrammar {
+        const tm = this.getCoreNodeModule("vscode-textmate", logger);
+        if (tm == null) { return undefined; }
+        const registry = new tm.Registry();
+        if (registry == null) { return undefined; }
+        const grammarPath = this.powerShellGrammarPath();
+        if (grammarPath == null) {
+            return undefined;
+        } else {
+            logger.writeDiagnostic(`PowerShell grammar file specified as ${grammarPath}`);
+        }
+        try {
+            return registry.loadGrammarFromPathSync(grammarPath);
+        } catch (err) {
+            logger.writeError(`Error while loading the PowerShell grammar file at ${grammarPath}`, err);
+        }
+    }
+
+    /**
+     * Returns a node module installed within VSCode, or null if it fails.
+     * Some node modules (e.g. vscode-textmate) cannot be required directly, instead the known module locations
+     * must be tried. Documented in https://github.com/Microsoft/vscode/issues/46281
+     * @param moduleName Name of the module to load e.g. vscode-textmate
+     * @param logger     The logging object to send messages to
+     * @returns          The required module, or null if the module cannot be required
+     */
+    private getCoreNodeModule(moduleName: string, logger: Logger) {
+        // Attempt to load the module from known locations
+        const loadLocations: string[] = [
+            `${vscode.env.appRoot}/node_modules.asar/${moduleName}`,
+            `${vscode.env.appRoot}/node_modules/${moduleName}`,
+        ];
+
+        for (const filename of loadLocations) {
+            try {
+                const mod = require(filename);
+                logger.writeDiagnostic(`Succesfully required ${filename}`);
+                return mod;
+            } catch (err) {
+                logger.writeError(`Error while attempting to require ${filename}`, err);
+            }
+        }
+        return null;
+    }
+
+     /**
+      * Search all of the loaded extenions for the PowerShell grammar file
+      * @returns The absolute path to the PowerShell grammar file.  Returns undefined if the path cannot be located.
+      */
+     private powerShellGrammarPath(): string {
+        try {
+            const psGrammars =
+                vscode.extensions.all
+                    // Filter out extensions without grammar contributions
+                    .filter((ext) => {
+                        return ext.packageJSON && ext.packageJSON.contributes && ext.packageJSON.contributes.grammars;
+                    })
+                    // For each extension, extract the list of grammar languages
+                    .reduce((grammarList: IExtensionGrammar[], ext) =>
```

Comment:
The `reduce` here still seems a bit complicated to me, given the spread operators and the internal `map` it uses. It's annoying because I love a good fold, but I think for maintainability's sake we should change it. Maybe to something like:

```typescript
// Go through all the extension packages and search for PowerShell grammars,
// returning the path to the first we find
foreach (const ext of vscode.extensions.all) {
    if (!(ext.packageJSON && ext.packageJSON.contributes && ext.packageJSON.contributes.grammars)) {
        continue;
    }

    foreach (const grammar: IGrammarExtension of (ext.packageJSON as IExtensionPackage).contributes.grammars) {
        if (grammar.language !== "powershell") {
            continue;
        }

        return path.join(ext.extensionPath, grammar.path);
    }
}

return undefined;
```

This way, we return when the first PowerShell grammar is found and there's no array copying with each accumulator pass.

(If I've totally misunderstood what this method does, just ignore me 😄)